p2p/host/basic/natmgr.go

```diff
@@ -65,7 +65,13 @@ func newNatManager(net network.Network) *natManager {
 func (nmgr *natManager) Close() error {
 	nmgr.ctxCancel()
 	nmgr.refCount.Wait()
-	return nil
+	var err error
+	nmgr.natMx.Lock()
+	if nmgr.nat != nil {
+		err = nmgr.nat.Close()
+	}
+	nmgr.natMx.Unlock()
```

Comment:
We should do this in a defer in `background()`. That way we can be sure it only happens once.

Reply:
Are you saying that someone could call Close many times? In my opinion this would be incorrect.

Reply:
People often `defer thing.Close()`. It should usually be idempotent.

Reply:
But doing this (closing it more than one) here should not be a problem, firstly `nmgr.nat.Close()` is idempotent, secondly by moving this to a deferred function in `background()` we lose the ability to return the error from `nmgr.nat.Close()`.

Reply:
Generally I agree that these child objects should be closed when the parent is closed directly.

However, the current code already follows a pattern of _background worker + context cancel for cleanup_ so this change should follow the conventions of the current code rather than introducing a new pattern. If that makes sense, @watjurk, could you please move this to a defer in the background instead?